File: api/src/main/java/net/blueshell/api/service/contribution/ContributionReminderService.java
package net.blueshell.api.service.contribution;

import net.blueshell.api.base.BaseModelService;
import net.blueshell.api.job.email.ContributionReminderEmailJob;
import net.blueshell.api.model.contribution.Contribution;
import net.blueshell.api.model.contribution.ContributionPeriod;
import net.blueshell.api.model.contribution.ContributionReminder;
import net.blueshell.api.repository.contribution.ContributionReminderRepository;
import net.blueshell.api.repository.contribution.ContributionRepository;
import net.blueshell.api.service.email.EmailService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.sql.Timestamp;
import java.time.Instant;
import java.util.HashSet;
import java.util.List;

@Service
public class ContributionReminderService extends BaseModelService<ContributionReminder, ContributionReminderRepository> {

    private final ContributionPeriodService periodService;
    private final EmailService emails;
    private final ContributionReminderEmailJob reminderEmailJob;

    @Autowired
    public ContributionReminderService(ContributionReminderRepository repository, EmailService emails, ContributionPeriodService periodService, ContributionReminderEmailJob reminderEmailJob) {
        super(repository);
        this.emails = emails;
        this.periodService = periodService;
        this.reminderEmailJob = reminderEmailJob;
    }

    @Transactional(readOnly = true)
    public List<ContributionReminder> findByContributionPeriodId(Long contributionPeriodId) {
        var contributionPeriod = periodService.findById(contributionPeriodId);
        return repository.findByContributionPeriod(contributionPeriod);
    }

    public void sendReminder(ContributionReminder reminder) {
        reminderEmailJob.send(reminder.getId());
    }

    public void sendReminders(List<ContributionReminder> reminders) {
        for (var reminder : reminders) {
            sendReminder(reminder);
        }
    }
}

File: api/src/main/kotlin/net/blueshell/api/service/contribution/ContributionReminderService.kt
package net.blueshell.api.service.contribution

import net.blueshell.api.common.event.job.ContributionReminderEmailEvent
import net.blueshell.api.model.contribution.ContributionReminder
import net.blueshell.api.repository.contribution.ContributionReminderRepository
import net.blueshell.api.service.base.BaseModelService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ContributionReminderService @Autowired constructor(
    repository: ContributionReminderRepository,
    private val periodService: ContributionPeriodService,
    private val eventPublisher: ApplicationEventPublisher
) : BaseModelService<ContributionReminder, ContributionReminder.Id, ContributionReminderRepository>(repository) {
    @Transactional(readOnly = true)
    fun findByContributionPeriodId(contributionPeriodId: Long): MutableList<ContributionReminder> {
        periodService.findById(contributionPeriodId)
        return repository.findById_ContributionPeriodId(contributionPeriodId)
    }

    fun sendReminder(reminder: ContributionReminder) {
        val reminderId = reminder.id
        eventPublisher.publishEvent(
            ContributionReminderEmailEvent(
                reminderId.userId,
                reminderId.contributionPeriodId
            )
        )
    }

    fun sendReminders(reminders: MutableList<ContributionReminder>) {
        for (reminder in reminders) {
            sendReminder(reminder)
        }
    }
}
